File: WordListBraceExpansion.py
# Time Complexity : O(m^n) where m is the m is the number of blocks and n is the length of the block 
# Space Complexity :O(2n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# Your code here along with comments explaining your approach

class Solution:
    result = None
    def expand(self, s):
        if len(s) == 0:
            return []
        self.result = []
        i = 0 
        words = []
       
       # O(n)
        while i < len(s):
            temp = []
            if s[i] == '{':
                i+=1 
                while s[i] != '}':
                    if s[i] != ',':
                        temp.append(s[i])
                    i += 1  
                words.append(temp)
            else:
                temp.append(s[i])
                words.append(temp)
            i+=1
        
        self.backtrack(words, 0, [])
        return self.result 
    
    def backtrack(self, words, indx, sb):
        # base 
        if indx == len(words):
            self.result.append(''.join(sb))
            return 
        #logic 
        for c in words[indx]:
            sb.append(c) 
            self.backtrack(words, indx+1, sb)
            sb.pop()
        
        return
                    

if __name__ =="__main__":
    s = Solution()
    # Input= "{a,b}c{d,e}f"
    # Output= ["acdf","acef","bcdf","bcef"]
    # res = s.expand(Input)
    # print(res)
    # assert res == Output
    
    input = 'abcd'
    output = ['abcd']
    res = s.expand(input)
    print(res)
    res == output
    

File: WordListBraceExpansion.java
//TC:O(N ^ M/N), M-> string length, N : no. of decisions at each level, M/N -> no. of groups
//SC:Exponential
// Did it run successfully on Leetcode? : Yes
class Solution {
    List<String> result;
    public String[] expand(String s) {
        if ( s == null || s.length() == 0)
            return new String[0];
        result = new ArrayList();
        List<List<Character>> blocks = new ArrayList();
        int i = 0;
        while ( i < s.length()){
            char c = s.charAt(i);
            List<Character> block = new ArrayList();
            if (c == '{'){
                i++;
                while (s.charAt(i) != '}'){
                    if (s.charAt(i) != ','){
                       block.add(s.charAt(i));
                    }
                    i++;
                }
            }
            else {
                block.add(c);   
            }
            Arrays.sort(block.toArray());
            blocks.add(block);
            i++;
        }
        backtrack(blocks, 0, new StringBuilder());
        String[] ans = new String[result.size()];
        for ( int j = 0; j < result.size(); j++)
        {
            ans[j] = result.get(j);
        }
        Arrays.sort(ans);
        return ans;
    }
    private void backtrack(List<List<Character>> blocks, int index, StringBuilder sb)
    {
        //base
        if (index == blocks.size()){
            result.add(sb.toString());
            return;
        }
        //logic
        List<Character> block = blocks.get(index);
        for ( int i = 0; i < block.size(); i++)
        {
            char c = block.get(i);
            //action
            sb.append(c);      
            //recurse
            backtrack(blocks, index+1, sb);
            //backtrack
            sb.setLength(sb.length() - 1);
        }
    }
}
